# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/platforms/android/RemoteDataDecoder.cpp
# Commit: 290917deea87
# Full Hash: 290917deea879f965f6f1ad441df945c41c01c14
# Author: Jamie Nicol <jnicol@mozilla.com>
# Date: 2023-12-13 21:42:49
# Regressor Bug: 1866020
# File Overlap Count: 1
# Description:
#   Bug 1866020 - Override buggy colorspace conversion on Pixel devices. r=gw,padenot,geckoview-reviewers,owlish
#   
#   Pixel 6, 7, and 8 devices running Android 14 are affected by a bug
#   where video frames with SMPTE 432 color primaries are rendered
#   incorrectly when sampled from an external texture in GLES. To work
# ==============================================================================

diff -r 12e266e296c1 -r 290917deea87 dom/media/platforms/android/RemoteDataDecoder.cpp
--- a/dom/media/platforms/android/RemoteDataDecoder.cpp	Wed Dec 13 08:29:00 2023 +0000
+++ b/dom/media/platforms/android/RemoteDataDecoder.cpp	Wed Dec 13 09:06:29 2023 +0000
@@ -67,6 +67,17 @@
   java::Sample::GlobalRef mSample;
 };
 
+static bool areSmpte432ColorPrimariesBuggy() {
+  if (jni::GetAPIVersion() >= 34) {
+    const auto socManufacturer =
+        java::sdk::Build::SOC_MANUFACTURER()->ToString();
+    if (socManufacturer.EqualsASCII("Google")) {
+      return true;
+    }
+  }
+  return false;
+}
+
 class RemoteVideoDecoder final : public RemoteDataDecoder {
  public:
   // Render the output to the surface when the frame is sent
@@ -402,9 +413,16 @@
     }
 
     if (ok && (size > 0 || presentationTimeUs >= 0)) {
+      // On certain devices SMPTE 432 color primaries are rendered incorrectly,
+      // so we force BT709 to be used instead. The magic number 10 comes from
+      // libstagefright's kColorStandardDCI_P3.
+      static bool isSmpte432Buggy = areSmpte432ColorPrimariesBuggy();
+      bool forceBT709ColorSpace = isSmpte432Buggy && mColorSpace == Some(10);
+
       RefPtr<layers::Image> img = new layers::SurfaceTextureImage(
           mSurfaceHandle, inputInfo.mImageSize, false /* NOT continuous */,
-          gl::OriginPos::BottomLeft, mConfig.HasAlpha(), mTransformOverride);
+          gl::OriginPos::BottomLeft, mConfig.HasAlpha(), forceBT709ColorSpace,
+          mTransformOverride);
       img->AsSurfaceTextureImage()->RegisterSetCurrentCallback(
           std::move(releaseSample));
 